### PR TITLE
WIP draft of synchronized output for parallel tasks

### PR DESCRIPTION
**Pull request by gdubicki**
Please check out how this would be used in https://github.com/dmerejkowsky/tsrc/pull/304 .

### CODE REVIEW

File: cli_ui/__init__.py [hunk 1]
```diff
@@ -234,14 +247,56 @@ def write_and_flush(fileobj: FileObj, to_write: str) -> None:
...
 
+lock = RLock()
+
+
+class Buffer(io.StringIO):
+    def __init__(self, is_like_tty: bool) -> None:
+        super().__init__()
+        self.is_like_tty = is_like_tty
+
+    def flush(self) -> None:
```
Comment (cgestes) [inline cli_ui/__init__.py:258]: Who could call this?

File: cli_ui/__init__.py [hunk 2]
```diff
@@ -57,6 +60,7 @@ def setup(
     color: str = "auto",
     title: str = "auto",
     timestamp: bool = False,
+    parallel: bool = False,
```
Comment (cgestes) [inline cli_ui/__init__.py:63]: Is this option needed? Seems like buffer="something" is needed to qctivate the feature anyway.so on/off for the default call seems to have no change in behavior.

File: cli_ui/__init__.py [hunk 3]
```diff
@@ -234,14 +247,56 @@ def write_and_flush(fileobj: FileObj, to_write: str) -> None:
...
+    def __init__(self, is_like_tty: bool) -> None:
+        super().__init__()
+        self.is_like_tty = is_like_tty
+
+    def flush(self) -> None:
+        # this is intentially no-op to prevent clearing the buffer before printing it
+        pass
+
+    def print(self) -> None:
+        with lock:
```
Comment (cgestes) [inline cli_ui/__init__.py:263]: This protect from printing 2 buffers interleaved?
Having buffers seems to allow delaying the printing to when it is wanted (or even make it conditional).
I wonder if this lock is aalway wanted. Also if its even needed does the underlying function locks? Or simply the GIL?

**Review by cgestes — COMMENTED** [review]
I am not a maintainer here. Just being curious.

Also another possible API could be:

b1 = ui.Buffer(toTTY=True)
b1.info("lalala")
b1.print()

File: cli_ui/__init__.py (hunk 2 above)
Comment (gdubicki) [reply cli_ui/__init__.py:63]: The idea behind this approach is that you can have the same calls for printing the output regardless if you are paralellizing your tasks or not, always passing some session_id. Only when the parallelization is enabled in the config though, the output is in fact printed out only on explicit `print_buffer` calls.

File: cli_ui/__init__.py (hunk 1 above)
Comment (gdubicki) [reply cli_ui/__init__.py:258]: It’s being called by some cli_ui methods and this noop implementation is to not have to change that when the output goes to the buffer.

File: cli_ui/__init__.py (hunk 3 above)
Comment (gdubicki) [reply cli_ui/__init__.py:263]: Yes, it’s for making sure that outputting the buffers in parallel threads will not make the output interleaved.

Of course we could make using the lock conditional but why would we want that? Can you give an example of a use case for it?

**Comment by gdubicki** [comment]
> I am not a maintainer here. Just being curious.
> 
Sure, thanks for a review and the questions. :)

> Also another possible API could be:
> 
> b1 = ui.Buffer(toTTY=True)
> b1.info("lalala")
> b1.print()

Yeah, but then you would have a lot of conditionals everywhere where the given code could be parallelized or not, depending on the settings.

Or you would have a worse user experience when it would not be parallelized as the output would show only at the end of the task, on the `print_buffer` call.

Right? Or am I missing something in your proposal?

I am not very happy with what I did here and am very open to listening to better ideas. :)

File: cli_ui/__init__.py (hunk 2 above)
Comment (gdubicki) [reply cli_ui/__init__.py:63]: Please check out how this would be used in another draft PR: https://github.com/dmerejkowsky/tsrc/pull/304

**Comment by dmerejkowsky** [comment]
So I've reviewed the code and I think we don't want that kind of feature inside `cli-ui` after all.

This work was started in the hope it will help fix output display when using `tsrc` with parallel tasks, but it turned out we only need to patch `tsrc`, not `cli-ui`.

Designing an API that works well when use concurrently is a hard problem, so I'd prefer leaving it out of `cli-ui`.

We can still patch the documentation to advise people willing to use `cli-ui` in a concurrent context they should use something like:

```python
stdout_lock = threading.RLock()

with stdout_lock:
    cli_ui.info(...)
```

But I think it's best to leave the responsibility of the lock on stdout in the hands of the user of the library.

**Comment by cgestes** [comment]
Hi,

I was wondering about how to solve this nicely for the author of the PR.

The API already take a fileobj as a parameter everywhere.

This fileobj can be customized to support locking, or to support grouped delayed write. 
I guess this fileobj could also be defined at setup time to get a global locking behavior.

Maybe some examples would make sense if this PR does not belong to python_cli_ui?

**Comment by dmerejkowsky** [comment]
> This fileobj can be customized to support locking, or to support grouped delayed write.

That's a good idea, but I'm not sure it would work ... It sure would make or a nice example :)

But again, having API users manage a stdout lock themselves seems much easier for everyone.

**Comment by dmerejkowsky** [comment]
Opened a separate issue there: #77

**Pull request closed without merging**